test/provider/AppWindow.unittest.ts

```diff
@@ -0,0 +1,164 @@
+        testAppWindow.removeContextListener();
+        expect(testAppWindow.hasContextListener()).toBe(false);
+    });
+
+    test('The state of one TestAppWindow does not affect another', () => {
+        const secondTestAppWindow = new TestAppWindow(fakeIdentity, fakeAppInfo, mockChannel, Date.now(), 0);
+        secondTestAppWindow.addContextListener();
+
+        expect(testAppWindow.hasContextListener()).toBe(false);
+    });
+});
+
+describe('When querying if a window is ready to receive contexts', () => {
+    let testAppWindow: TestAppWindow;
+
+    beforeEach(() => {
+        // All tests in this section will use fake timers to allow us to control the Promise races precisely
+        jest.useFakeTimers();
+
+        testAppWindow = new TestAppWindow(fakeIdentity, fakeAppInfo, mockChannel, Date.now(), 0);
+    });
+
+    afterEach(() => {
+        jest.useRealTimers();
+    });
+
+    test('A window with a context listener already registered returns true immediately', async () => {
+        testAppWindow.addContextListener();
+
+        // Use a jest spy to track the timing of when the promise resolves without awaiting
+        const timingSpy = jest.fn();
+        testAppWindow.isReadyToReceiveContext().then(timingSpy);
+
+        // Do not advance time, but let any pending promises be actioned
+        await Promise.resolve();
+
+        // Promise should have resolved immediately, so spy should have been invoked
+        expect(timingSpy).toHaveBeenCalledWith(true);
+    });
+
+    describe('When the window does not have a listener registered', () => {
+        test('If the window was created longer than the timeout in the past, the promise resolves false immediately', async () => {
+            // @ts-ignore Updating the creation time to test old windows
+            testAppWindow._creationTime = Date.now() - 10000;
+
+            // Use a jest spy to track the timing of when the promise resolves without awaiting
+            const timingSpy = jest.fn();
+            testAppWindow.isReadyToReceiveContext().then(timingSpy);
+
+            // Do not advance time, but let any pending promises be actioned
+            await Promise.resolve();
+
+            // Promise should have resolved immediately, so spy should have been invoked
+            expect(timingSpy).toHaveBeenCalledWith(false);
+        });
+
+        test('If the window registers a listener within the timeout, the promise resolves true shortly after registration', async () => {
+            // Use a jest spy to track the timing of when the promise resolves without awaiting
+            const timingSpy = jest.fn();
+            testAppWindow.isReadyToReceiveContext().then(timingSpy);
+
+            jest.advanceTimersByTime(1000);
+            await Promise.resolve();
+            expect(timingSpy).not.toHaveBeenCalled();
+
+            testAppWindow.addContextListener();
+
+            await Promise.resolve();
+            jest.advanceTimersByTime(5000);
```

Comment:
Copy over the fake time code form notification, which wraps together advancing time and resolving promises?

Reply:
Ah, didn't know we had that. Will take a look.

Reply:
There's a couple of leftover instances of `await Promise.resolve();` - one that can be replaced with `resolvePromiseChain`, and one that's commented one so I assume can just be deleted

Reply:
There's still line 107 using the raw `await Promise.resolve();`

Reply:
Ok, that should be fixed.

Reply:
And I searched through for any others and didn't see any.